### PR TITLE
FEATURE: Allow linking to content nodes directly

### PR DESCRIPTION
**Pull request by alexander-nitsche**
Instead of copying and pasting a generated uri of a content node it is supported to set a link to the content node directly. This is less error prone as it reflects changes of the anchor id automatically and allows to scan comfortably for missing content nodes.

TODOS:
- Check if the aspect implementation is state-of-the-art: It is heavily inspired by Neos' [PluginUriAspect](https://github.com/neos/neos-development-collection/blob/7.3/Neos.Neos/Classes/Aspects/PluginUriAspect.php).
- Check backwards compatibility for Neos < 7.

### CODE REVIEW

**Comment by alexander-nitsche** [comment]
> TODO: Document how-to list only content nodes with sectionId property set in link search dialog.

Cannot find a simple way to list only content nodes that have the sectionId property set to a non-empty string. Falling back to allowing all nodes that inherit from the `CodeQ.JumpMarkers:Mixin.SectionConfiguration` mixin.

**Comment by alexander-nitsche** [comment]
Hi @rolandschuetz , maybe it is a good point in time to streamline the wording a bit. There is a lot of words flowing around that describe the same:

- jump marker
- anchor id
- section id
- url fragment

What do you think of SectionId, SectionTitle and SectionMenu? **Section** is used in the Neos' UriBuilder for the Uri fragment part and for the node property sectionId already. I would replace any mentioning of **JumpMarker** as well as **Fragment** and **Anchor**. This can be organised in a separate issue of course.

**Comment by alexander-nitsche** [comment]
> TODOS:
>
>    Check if the aspect implementation is state-of-the-art: It is heavily inspired by Neos' [PluginUriAspect](https://github.com/neos/neos-development-collection/blob/7.3/Neos.Neos/Classes/Aspects/PluginUriAspect.php).
>    Check backwards compatibility for Neos < 7.

The implementation of the ContentNodeUriAspect class seems to be not compatible with Neos 4 and 5 yet, as working with the `$arguments['controllerArguments']` seems to have been introduced in Neos 7 earliest:

```
public function generateUriForNode(JoinPointInterface $joinPoint, NodeInterface $pageNode, NodeInterface $contentNode)
{
    ..

    // generate the uri for the page node with the content node as uri fragment
    $arguments['controllerArguments']['node'] = $pageNode;
    $joinPoint->setMethodArgument('controllerArguments', $arguments['controllerArguments']);
    $joinPoint->getProxy()->setSection($contentNode->getProperty('sectionId'));

    ..

    // restore the content node
    $arguments['controllerArguments']['node'] = $contentNode;
    $joinPoint->setMethodArgument('controllerArguments', $arguments['controllerArguments']);
    $joinPoint->getProxy()->setSection($originalSection);

    ..
}
```

Maybe you know with deeper Neos knowledge how to achive the same with being compatible with Neos 4, 5 and 7 at the same time?

File: README.md [hunk 1]
```diff
@@ -27,7 +27,35 @@ E.g. this code adds such ability to every Content NodeType:
...
+    position: 200
+    inlineEditable: true
+  properties:
+    text:
+      options:
+        ui:
+          inline:
+            editorOptions:
+              linking:
+                nodeTypes: ['Neos.Neos:Document', 'CodeQ.JumpMarkers:Mixin.SectionConfiguration']
```
Comment (rolandschuetz) [inline README.md]: There's a bug with `options` and I would add two proper examples here:
```
  properties:
    text:
      type: string
      defaultValue: ''
      ui:
        inlineEditable: true
        inline:
          editorOptions:
            linking:
              nodeTypes: ['Neos.Neos:Document', 'CodeQ.JumpMarkers:Mixin.SectionConfiguration']
    link:
      type: string
      ui:
        label: 'Link'
        reloadPageIfChanged: false
        inspector:
          group: general
          editor: 'Neos.Neos/Inspector/Editors/LinkEditor'
          editorOptions:
            linking:
              nodeTypes: ['Neos.Neos:Document', 'CodeQ.JumpMarkers:Mixin.SectionConfiguration']

File: README.md [hunk 2]
```diff
@@ -27,7 +27,35 @@ E.g. this code adds such ability to every Content NodeType:
...
+```
+
+Note that this allows any content node to be selected as a link target, including those that have not yet been assigned 
+an anchor id and therefore cannot be linked properly in the frontend.
```
Comment (rolandschuetz) [inline README.md:69]: "including those that have not yet been assigned an anchor id and therefore cannot be linked properly in the frontend."
That's a big weird, I would recommend that we have a fallback to always render ids with node.identifier like this:
```
prototype(CodeQ.JumpMarkers:NodeAnchorId) < prototype(Neos.Fusion:Component) {
    ...

    renderer = Neos.Fusion:Value {
        ...

        fallback = ${q(props.node).property('jumpMarkerTitle') || ('node-' + node.identifier)}
        ...
    }
}

```

File: Resources/Private/Translations/de/NodeTypes/Mixin/SectionConfiguration.xlf [hunk 3]
```diff
@@ -3,36 +3,36 @@
...
-				<source>Enter a section ID of the type short-text-without-special-characters. After you click Save, a button for copying the link appears below it. This link can be pasted in the text editor or in fields of the Inspector and will lead directly to the section.</source>
-				<target>Gib eine Abschnitts-ID in der Art kurzer-text-ohne-sonderzeichen ein. Nachdem du auf Speichern klickst, erscheint darunter ein Button zum Kopieren des Links. Dieser Link im Texteditor oder in Feldern des Inspektors eingefügt werden und führt direkt zum Abschnitt.</target>
+				<source>Enter an URL fragment of the type short-text-without-special-characters. After you click Save, a button for copying the link appears below it. This link can be pasted in the text editor or in fields of the Inspector and will lead directly to the section.</source>
+				<target>Geben Sie ein URL-Fragment vom Typ kurztext-ohne-sonderzeichen ein. Nachdem Du auf Speichern geklickt hast, erscheint darunter eine Schaltfläche zum Kopieren des Links. Dieser Link kann in den Texteditor oder in Felder des Inspektors eingefügt werden und führt direkt zu dem Abschnitt.</target>
```
Comment (rolandschuetz) [inline Resources/Private/Translations/de/NodeTypes/Mixin/SectionConfiguration.xlf]: Wie sonst auch mit `Du`, ich würde das wie folgt umschreiben:

Gib hier ein Anker-URL-Fragment für in der Form kurztext-ohne-sonderzeichen ein. Nachdem Du auf Speichern geklickt hast, erscheint darunter eine Schaltfläche zum Kopieren des Links. Dieser Link kann in den Texteditor oder in Felder des Inspektors eingefügt werden und führt direkt zu dem Abschnitt.

File: Classes/Aspects/ContentNodeUriAspect.php [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+     */
+    public function generateUriForNode(JoinPointInterface $joinPoint, NodeInterface $pageNode, NodeInterface $contentNode)
+    {
+        /** @var UriBuilder $uriBuilder */
+        $arguments = $joinPoint->getMethodArguments();
+        $uriBuilder = $joinPoint->getProxy();
+        $originalSection = $uriBuilder->getSection();
+
+        // generate the uri for the page node with the content node as uri fragment
+        $arguments['controllerArguments']['node'] = $pageNode;
```
Comment (rolandschuetz) [inline Classes/Aspects/ContentNodeUriAspect.php:67]: Let's just raise the minimal requirement for Neos to 7.0.

**Review by rolandschuetz — CHANGES_REQUESTED** [review]
Hello,

I really like this feature. I have tested it and it worked well. Furthermore, I have added a few minor things. I already really like your wording. 

I think with this minor changes we can release this as a Version 3

**Comment by alexander-nitsche** [comment]
Rebased the branch on the current master and added two commits.

File: README.md (hunk 1 above)
Comment (alexander-nitsche) [reply README.md]: Done.

File: Resources/Private/Translations/de/NodeTypes/Mixin/SectionConfiguration.xlf (hunk 3 above)
Comment (alexander-nitsche) [reply Resources/Private/Translations/de/NodeTypes/Mixin/SectionConfiguration.xlf]: Adjusted, but slightly changed. Done.

File: Classes/Aspects/ContentNodeUriAspect.php (hunk 4 above)
Comment (alexander-nitsche) [reply Classes/Aspects/ContentNodeUriAspect.php:67]: I would leave this to you, as I have no experience with releasing packages for specific Neos versions and how to handle these flow migrations in composer.json in particular.

File: README.md (hunk 2 above)
Comment (alexander-nitsche) [reply README.md:69]: I am not sure here. Currently a section id or section title has to be actively set by the editor to have some URL with fragment for further use. With this fallback we would switch from opt-in to always-in. Maybe we should add a global switch to turn this feature on and off. Do we really want to show the node id to the outside world?

Comment (rolandschuetz) [reply README.md:69]: I wrote an aspect handling this, not only nodes with a jumpMarkerTitle or sectionId id are shown.

**Pull request merged**